File: src/accessiweather/display/presentation/environmental.py
"""Presentation helpers for environmental data such as air quality."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime

from ...models import AppSettings, EnvironmentalConditions, HourlyAirQuality, Location
from .formatters import format_display_datetime

_AIR_QUALITY_GUIDANCE: dict[str, str] = {
    "Good": "Air quality is satisfactory; enjoy normal outdoor activities.",
    "Moderate": (
        "Air quality is acceptable. People unusually sensitive to air pollution should"
        " reduce prolonged or heavy outdoor exertion."
    ),
    "Unhealthy for Sensitive Groups": (
        "Sensitive groups (people with lung or heart disease, older adults, children)"
        " should reduce prolonged or heavy outdoor exertion."
    ),
    "Unhealthy": (
        "Everyone should reduce prolonged or heavy outdoor exertion; sensitive groups"
        " should avoid extended time outdoors."
    ),
    "Very Unhealthy": "Avoid outdoor exertion and move activities indoors when possible.",
    "Hazardous": "Avoid all outdoor activity and follow local emergency air quality guidance.",
}

_DEFAULT_GUIDANCE = "Monitor local guidance and limit exposure if you notice symptoms."

_UV_INDEX_GUIDANCE: dict[str, str] = {
    "Low": "No protection needed. You can safely stay outside.",
    "Moderate": (
        "Take precautions: wear sunscreen SPF 30+, a hat, and sunglasses."
        " Seek shade during midday hours when the sun is strongest."
    ),
    "High": (
        "Protection essential: apply SPF 30+ sunscreen every 2 hours, wear protective"
        " clothing, a wide-brimmed hat, and UV-blocking sunglasses. Seek shade during midday."
    ),
    "Very High": (
        "Extra protection required: use SPF 50+ sunscreen, wear long sleeves and pants,"
        " a wide-brimmed hat, and sunglasses. Minimize sun exposure between 10 AM and 4 PM."
    ),
    "Extreme": (
        "Take all precautions: avoid sun exposure between 10 AM and 4 PM if possible."
        " Use SPF 50+ sunscreen, wear full protective clothing, and stay in shade when outdoors."
    ),
}

_POLLUTANT_LABELS: dict[str, str] = {
    "PM2_5": "PM2.5",
    "PM10": "PM10",
    "O3": "Ozone",
    "SO2": "Sulfur Dioxide",
    "NO2": "Nitrogen Dioxide",
    "CO": "Carbon Monoxide",
}


@dataclass(slots=True)
class AirQualityPresentation:
    """Structured representation of air quality metrics."""

    title: str
    summary: str
    guidance: str | None = None
    details: list[str] = field(default_factory=list)
    fallback_text: str = ""
    updated_at: str | None = None
    sources: list[str] = field(default_factory=list)


def build_air_quality_panel(
    location: Location,
    environmental: EnvironmentalConditions,
    settings: AppSettings | None = None,
) -> AirQualityPresentation | None:
    """Create an accessible presentation of air quality metrics."""
    index = environmental.air_quality_index
    pollutant = environmental.air_quality_pollutant
    category = environmental.air_quality_category

    if index is None and not category and not pollutant:
        return None

    summary_line = _build_summary_line(index, category)
    pollutant_line = _build_pollutant_line(pollutant)
    updated_line = _build_updated_line(environmental.updated_at, settings)
    guidance = _AIR_QUALITY_GUIDANCE.get(category or "", _DEFAULT_GUIDANCE)
    sources = sorted({source for source in environmental.sources if source})

    lines = [f"Air quality for {location.name}"]
    details: list[str] = []

    # Pollen data
    pollen_line = _build_pollen_line(environmental)
    if pollen_line:
        lines.append(f"• {pollen_line}")
        details.append(pollen_line)

        # Add detailed pollen breakdown if available
        pollen_details = format_pollen_details(environmental)
        if pollen_details:
            details.append(pollen_details)

    if summary_line:
        lines.append(f"• {summary_line}")
        details.append(summary_line)
    if pollutant_line:
        lines.append(f"• {pollutant_line}")
        details.append(pollutant_line)
    if updated_line:
        lines.append(f"• {updated_line}")
        details.append(updated_line)
    if guidance:
        lines.append(f"• Advice: {guidance}")

    if sources:
        lines.append(f"Sources: {', '.join(sources)}")

    fallback_text = "\n".join(lines)

    summary_value = summary_line or "Air quality data not available."
    if pollutant_line:
        summary_value = f"{summary_value} – {pollutant_line}"
    if pollen_line and not summary_line:
        summary_value = pollen_line
    elif pollen_line:
        summary_value = f"{summary_value}. {pollen_line}"

    return AirQualityPresentation(
        title=f"Air quality for {location.name}",
        summary=summary_value,
        guidance=guidance,
        details=details,
        fallback_text=fallback_text,
        updated_at=updated_line,
        sources=sources,
    )


def _build_summary_line(index: float | None, category: str | None) -> str | None:
    if index is None and not category:
        return None

    parts: list[str] = []
    if index is not None:
        rounded = int(round(index))
        parts.append(f"AQI {rounded}")
    if category:
        category_text = category.strip()
        parts.append(f"({category_text})" if parts else category_text)

    return " ".join(parts).strip() if parts else None


def _build_pollutant_line(pollutant: str | None) -> str | None:
    if not pollutant:
        return None

    label = pollutant.strip().upper()
    pretty = _POLLUTANT_LABELS.get(label)
    if pretty is None:
        pretty = label.replace("_", " ").title() if "_" in label else label
    return f"Dominant pollutant: {pretty}"


def _build_pollen_line(environmental: EnvironmentalConditions) -> str | None:
    """Build a summary line for pollen conditions."""
    if environmental.pollen_index is None and not environmental.pollen_category:
        return None

    parts: list[str] = []
    if environmental.pollen_category:
        parts.append(f"Pollen: {environmental.pollen_category}")
    elif environmental.pollen_index is not None:
        parts.append(f"Pollen Index: {int(round(environmental.pollen_index))}")

    if environmental.pollen_primary_allergen:
        parts.append(f"({environmental.pollen_primary_allergen})")

    return " ".join(parts) if parts else None


def _build_updated_line(
    updated_at: datetime | None, settings: AppSettings | None = None
) -> str | None:
    if not updated_at:
        return None

    if settings:
        time_display_mode = getattr(settings, "time_display_mode", "local")
        time_format_12hour = getattr(settings, "time_format_12hour", True)
        show_timezone_suffix = getattr(settings, "show_timezone_suffix", False)
    else:
        time_display_mode = "local"
        time_format_12hour = True
        show_timezone_suffix = False

    timestamp = format_display_datetime(
        updated_at,
        time_display_mode=time_display_mode,
        use_12hour=time_format_12hour,
        show_timezone=show_timezone_suffix,
    )
    return f"Updated {timestamp}"


def format_hourly_air_quality(
    hourly_data: list[HourlyAirQuality],
    settings: AppSettings | None = None,
    max_hours: int = 24,
) -> str | None:
    """
    Format hourly air quality forecast into readable text.

    Args:
        hourly_data: List of hourly air quality forecasts.
        settings: App settings for time formatting.
        max_hours: Maximum number of hours to include.

    Returns:
        Formatted string describing the hourly forecast, or None if no data.

    """
    if not hourly_data:
        return None

    time_format_12hour = getattr(settings, "time_format_12hour", True) if settings else True
    data = hourly_data[:max_hours]

    current = data[0]
    peak = max(data, key=lambda h: h.aqi)
    best = min(data, key=lambda h: h.aqi)

    lines = []
    lines.append(f"Current: AQI {current.aqi} ({current.category})")

    if len(data) >= 3:
        trend_start = data[0].aqi
        trend_end = data[2].aqi
        diff = trend_end - trend_start

        if diff > 20:
            lines.append(f"Trend: Worsening (AQI {trend_start} → {trend_end})")
        elif diff < -20:
            lines.append(f"Trend: Improving (AQI {trend_start} → {trend_end})")
        else:
            lines.append("Trend: Stable")

    if peak.aqi != current.aqi:
        peak_time = _format_time(peak.timestamp, time_format_12hour)
        lines.append(f"Peak: AQI {peak.aqi} ({peak.category}) at {peak_time}")

    if best.aqi < 100 and best.aqi != current.aqi:
        best_time = _format_time(best.timestamp, time_format_12hour)
        lines.append(f"Best time: AQI {best.aqi} at {best_time}")

    # Add individual hourly entries (show up to 12 hours)
    hourly_entries = data[:12]
    if hourly_entries:
        lines.append("")
        lines.append("Hourly Forecast:")
        for entry in hourly_entries:
            entry_time = _format_time(entry.timestamp, time_format_12hour)
            lines.append(f"  {entry_time}: AQI {entry.aqi} ({entry.category})")

    return "\n".join(lines)


def _format_time(dt: datetime, use_12hour: bool = True) -> str:
    """Format a datetime as a simple time string."""
    if use_12hour:
        return dt.strftime("%I:%M %p").lstrip("0")
    return dt.strftime("%H:%M")


def format_air_quality_summary(
    environmental: EnvironmentalConditions,
    settings: AppSettings | None = None,
) -> str:
    """
    Format current air quality as a summary string for the dialog.

    This is a pure function that accepts data and settings and returns formatted text.

    Args:
        environmental: Environmental conditions data containing AQI info.
        settings: App settings for time formatting preferences.

    Returns:
        Formatted summary string with AQI, category, pollutant, guidance, and timestamp.

    """
    lines: list[str] = []

    aqi = environmental.air_quality_index
    category = environmental.air_quality_category

    if aqi is not None or category:
        aqi_text = ""
        if aqi is not None:
            aqi_text = f"AQI: {int(round(aqi))}"
        if category:
            if aqi_text:
                aqi_text += f" ({category})"
            else:
                aqi_text = category
        lines.append(aqi_text)

    pollutant = environmental.air_quality_pollutant
    if pollutant:
        pollutant_name = _get_pollutant_display_name(pollutant)
        lines.append(f"Dominant pollutant: {pollutant_name}")

    guidance = _AIR_QUALITY_GUIDANCE.get(category or "", _DEFAULT_GUIDANCE)
    lines.append(f"Health guidance: {guidance}")

    if environmental.updated_at:
        time_format_12hour = getattr(settings, "time_format_12hour", True) if settings else True
        if time_format_12hour:
            timestamp = environmental.updated_at.strftime("%I:%M %p").lstrip("0")
        else:
            timestamp = environmental.updated_at.strftime("%H:%M")
        date_str = environmental.updated_at.strftime("%B %d, %Y")
        lines.append(f"Last updated: {timestamp} on {date_str}")

    return "\n".join(lines) if lines else "Air quality data not available."


def format_pollutant_details(
    hourly_data: list[HourlyAirQuality],
    dominant_pollutant: str | None = None,
) -> str:
    """
    Format pollutant measurements into readable text.

    This is a pure function that accepts data and returns formatted text.

    Args:
        hourly_data: List of hourly air quality data (uses first entry for current values).
        dominant_pollutant: Code of the dominant pollutant to highlight.

    Returns:
        Formatted pollutant breakdown with human-readable names and dominant indicator.

    """
    if not hourly_data:
        return "No pollutant data available."

    current = hourly_data[0]
    dominant_code = dominant_pollutant.strip().upper() if dominant_pollutant else None

    pollutants = [
        ("PM2_5", current.pm2_5, "µg/m³"),
        ("PM10", current.pm10, "µg/m³"),
        ("O3", current.ozone, "µg/m³"),
        ("NO2", current.nitrogen_dioxide, "µg/m³"),
        ("SO2", current.sulphur_dioxide, "µg/m³"),
        ("CO", current.carbon_monoxide, "µg/m³"),
    ]

    lines: list[str] = []
    for code, value, unit in pollutants:
        if value is not None:
            name = _POLLUTANT_LABELS.get(code, code)
            line = f"{name}: {value:.1f} {unit}"
            if dominant_code and code == dominant_code:
                line += " (dominant)"
            lines.append(line)

    return "\n".join(lines) if lines else "No pollutant measurements available."


def format_pollen_details(environmental: EnvironmentalConditions) -> str | None:
    """
    Format pollen measurements into readable text.

    Args:
        environmental: Environmental conditions data.

    Returns:
        Formatted pollen breakdown or None if no data.

    """
    items = []
    if environmental.pollen_tree_index is not None:
        items.append(f"Tree: {int(round(environmental.pollen_tree_index))}")
    if environmental.pollen_grass_index is not None:
        items.append(f"Grass: {int(round(environmental.pollen_grass_index))}")
    if environmental.pollen_weed_index is not None:
        items.append(f"Weed: {int(round(environmental.pollen_weed_index))}")

    if not items:
        return None

    return "Pollen Levels: " + ", ".join(items)


def _get_pollutant_display_name(pollutant_code: str) -> str:
    """Convert pollutant code to human-readable display name."""
    code = pollutant_code.strip().upper()
    if code in _POLLUTANT_LABELS:
        return _POLLUTANT_LABELS[code]
    if code == "PM2.5":
        return "PM2.5"
    if code == "OZONE":
        return "Ozone"
    return code.replace("_", " ").title() if "_" in code else code


def format_air_quality_brief(
    environmental: EnvironmentalConditions,
    settings: AppSettings | None = None,
) -> str:
    """
    Format a brief air quality summary for Current Conditions section.

    This is a pure function that returns a concise one-line summary with
    AQI value, category, and trend only. No detailed hourly breakdown.

    Args:
        environmental: Environmental conditions data containing AQI info.
        settings: App settings (unused, kept for API consistency).

    Returns:
        Brief summary like "AQI: 45 (Good) - Stable" or empty string if no data.

    """
    parts: list[str] = []

    aqi = environmental.air_quality_index
    category = environmental.air_quality_category

    if aqi is not None:
        aqi_text = f"AQI: {int(round(aqi))}"
        if category:
            aqi_text += f" ({category})"
        parts.append(aqi_text)
    elif category:
        parts.append(category)

    hourly_data = environmental.hourly_air_quality
    if hourly_data and len(hourly_data) >= 3:
        trend_start = hourly_data[0].aqi
        trend_end = hourly_data[2].aqi
        diff = trend_end - trend_start

        if diff > 20:
            parts.append("Worsening")
        elif diff < -20:
            parts.append("Improving")
        else:
            parts.append("Stable")
    elif parts:
        parts.append("Stable")

    return " - ".join(parts) if parts else ""

File: rust/crates/aw-core/src/display/environmental.rs
//! Air quality and pollen presentation.
//!
//! Port of the parts of `display/presentation/environmental.py` the app
//! uses: the air quality panel, pollen details and the UV category helper.
//! (`format_air_quality_summary`, `format_pollutant_details`,
//! `format_air_quality_brief` and the `environmental_hourly.py` formatters
//! are never called by the Python app.)

use chrono_tz::Tz;

use crate::display::models::AirQualityPresentation;
use crate::display::pyfmt::{round_int, title};
use crate::display::time::{format_display_datetime, PyDateTime, DEFAULT_DATE_FORMAT};
use crate::model::EnvironmentalConditions;
use crate::settings::AppSettings;

fn air_quality_guidance(category: &str) -> &'static str {
    match category {
        "Good" => "Air quality is satisfactory; enjoy normal outdoor activities.",
        "Moderate" => {
            "Air quality is acceptable. People unusually sensitive to air pollution should \
             reduce prolonged or heavy outdoor exertion."
        }
        "Unhealthy for Sensitive Groups" => {
            "Sensitive groups (people with lung or heart disease, older adults, children) \
             should reduce prolonged or heavy outdoor exertion."
        }
        "Unhealthy" => {
            "Everyone should reduce prolonged or heavy outdoor exertion; sensitive groups \
             should avoid extended time outdoors."
        }
        "Very Unhealthy" => "Avoid outdoor exertion and move activities indoors when possible.",
        "Hazardous" => "Avoid all outdoor activity and follow local emergency air quality guidance.",
        _ => "Monitor local guidance and limit exposure if you notice symptoms.",
    }
}

fn pollutant_label(code: &str) -> Option<&'static str> {
    Some(match code {
        "PM2_5" => "PM2.5",
        "PM10" => "PM10",
        "O3" | "OZONE" => "Ozone",
        "SO2" => "Sulfur Dioxide",
        "NO2" => "Nitrogen Dioxide",
        "CO" => "Carbon Monoxide",
        _ => return None,
    })
}

/// `_get_uv_category`: EPA/WHO UV category.
pub fn uv_category(uv_index: Option<f64>) -> Option<&'static str> {
    let uv = uv_index?;
    Some(if uv <= 2.0 {
        "Low"
    } else if uv <= 5.0 {
        "Moderate"
    } else if uv <= 7.0 {
        "High"
    } else if uv <= 10.0 {
        "Very High"
    } else {
        "Extreme"
    })
}

fn nonempty(s: &Option<String>) -> Option<&str> {
    s.as_deref().filter(|s| !s.is_empty())
}

/// `build_air_quality_panel`.
pub fn build_air_quality_panel(
    location_name: &str,
    environmental: &EnvironmentalConditions,
    settings: &AppSettings,
    location_zone: Option<Tz>,
) -> Option<AirQualityPresentation> {
    let index = environmental.air_quality_index;
    let pollutant = nonempty(&environmental.air_quality_pollutant);
    let category = nonempty(&environmental.air_quality_category);
    if index.is_none() && category.is_none() && pollutant.is_none() {
        return None;
    }

    // Python only ever tests this line for truthiness, so "" acts like None.
    let summary_line = summary_line(index, category).filter(|s| !s.is_empty());
    let pollutant_line = pollutant.map(pollutant_line);
    let updated_line = environmental.updated_at.map(|t| {
        let stamp = format_display_datetime(
            &PyDateTime::aware(t, location_zone),
            &settings.time_display_mode,
            settings.time_format_12hour,
            settings.show_timezone_suffix,
            DEFAULT_DATE_FORMAT,
        );
        format!("Updated {stamp}")
    });
    let guidance = air_quality_guidance(category.unwrap_or(""));
    let mut sources: Vec<String> = environmental
        .sources
        .iter()
        .filter(|s| !s.is_empty())
        .cloned()
        .collect();
    sources.sort();
    sources.dedup();

    let title = format!("Air quality for {location_name}");
    let mut lines = vec![title.clone()];
    let mut details = Vec::new();

    let pollen_line = pollen_line(environmental);
    if let Some(p) = &pollen_line {
        lines.push(format!("• {p}"));
        details.push(p.clone());
        if let Some(pd) = format_pollen_details(environmental) {
            details.push(pd);
        }
    }
    for line in [&summary_line, &pollutant_line, &updated_line]
        .into_iter()
        .flatten()
    {
        lines.push(format!("• {line}"));
        details.push(line.clone());
    }
    lines.push(format!("• Advice: {guidance}"));
    if !sources.is_empty() {
        lines.push(format!("Sources: {}", sources.join(", ")));
    }

    let mut summary = summary_line
        .clone()
        .unwrap_or_else(|| "Air quality data not available.".into());
    if let Some(p) = &pollutant_line {
        summary = format!("{summary} – {p}");
    }
    if let Some(p) = &pollen_line {
        summary = if summary_line.is_none() {
            p.clone()
        } else {
            format!("{summary}. {p}")
        };
    }

    Some(AirQualityPresentation {
        title,
        summary,
        guidance: Some(guidance.into()),
        details,
        fallback_text: lines.join("\n"),
        updated_at: updated_line,
        sources,
    })
}

/// `_build_summary_line`.
fn summary_line(index: Option<f64>, category: Option<&str>) -> Option<String> {
    if index.is_none() && category.is_none() {
        return None;
    }
    let mut parts: Vec<String> = Vec::new();
    if let Some(i) = index {
        parts.push(format!("AQI {}", round_int(i)));
    }
    if let Some(c) = category {
        let c = c.trim();
        parts.push(if parts.is_empty() {
            c.to_string()
        } else {
            format!("({c})")
        });
    }
    Some(parts.join(" ").trim().to_string())
}

/// `_build_pollutant_line`.
fn pollutant_line(pollutant: &str) -> String {
    let label = pollutant.trim().to_uppercase();
    let pretty = match pollutant_label(&label) {
        Some(p) => p.to_string(),
        None if label.contains('_') => title(&label.replace('_', " ")),
        None => label,
    };
    format!("Dominant pollutant: {pretty}")
}

/// `_build_pollen_line`.
fn pollen_line(env: &EnvironmentalConditions) -> Option<String> {
    let category = nonempty(&env.pollen_category);
    if env.pollen_index.is_none() && category.is_none() {
        return None;
    }
    let mut parts: Vec<String> = Vec::new();
    if let Some(c) = category {
        parts.push(format!("Pollen: {c}"));
    } else if let Some(i) = env.pollen_index {
        parts.push(format!("Pollen Index: {}", round_int(i)));
    }
    if let Some(a) = nonempty(&env.pollen_primary_allergen) {
        parts.push(format!("({a})"));
    }
    (!parts.is_empty()).then(|| parts.join(" "))
}

/// `format_pollen_details`: "Pollen Levels: Tree: 3, Grass: 1".
pub fn format_pollen_details(env: &EnvironmentalConditions) -> Option<String> {
    let items: Vec<String> = [
        ("Tree", env.pollen_tree_index),
        ("Grass", env.pollen_grass_index),
        ("Weed", env.pollen_weed_index),
    ]
    .into_iter()
    .filter_map(|(name, v)| v.map(|v| format!("{name}: {}", round_int(v))))
    .collect();
    (!items.is_empty()).then(|| format!("Pollen Levels: {}", items.join(", ")))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn panel_combines_pollen_and_air_quality() {
        let env = EnvironmentalConditions {
            air_quality_index: Some(52.5),
            air_quality_category: Some("Moderate".into()),
            air_quality_pollutant: Some("pm2_5".into()),
            pollen_category: Some("High".into()),
            pollen_primary_allergen: Some("Oak".into()),
            pollen_tree_index: Some(4.5),
            sources: vec!["Open-Meteo".into(), "".into(), "AirNow".into(), "AirNow".into()],
            ..Default::default()
        };
        let p = build_air_quality_panel("Home", &env, &AppSettings::default(), None).unwrap();
        assert_eq!(p.summary, "AQI 52 (Moderate) – Dominant pollutant: PM2.5. Pollen: High (Oak)");
        assert_eq!(p.sources, ["AirNow", "Open-Meteo"]);
        assert_eq!(p.details[1], "Pollen Levels: Tree: 4");
        assert!(p.fallback_text.ends_with("Sources: AirNow, Open-Meteo"));
    }

    #[test]
    fn unknown_pollutant_codes_are_title_cased() {
        assert_eq!(pollutant_line("nh3_x"), "Dominant pollutant: Nh3 X");
        assert_eq!(pollutant_line("pm10"), "Dominant pollutant: PM10");
        assert_eq!(uv_category(Some(2.5)), Some("Moderate"));
    }
}
